src/TeamSyncer.ts

```diff
@@ -0,0 +1,317 @@
+            }
+            if (channels.whitelist) {
+                if (!channels.whitelist.includes(itemId)) {
+                    log.warn("Not bridging channel, not in whitelist");
+                    return false;
+                }
+            }
+        }
+        // More granular please.
+        return true;
+    }
+
+    public async onChannelAdded(teamId: string, channelId: string, name: string, creator: string) {
+        // Should create the channel?
+        log.info(`${teamId} created channel ${channelId}`);
+        const existingChannel = this.main.rooms.getBySlackChannelId(channelId);
+        if (existingChannel) {
+            log.info("Channel already exists in datastore, not bridging");
+            return;
+        }
+        const client = await this.main.clientFactory.getTeamClient(teamId);
+        const { channel } = (await client.conversations.info({ channel: channelId })) as ConversationsInfoResponse;
+        if (!channel.is_channel || channel.is_private) {
+            log.warn("Not creating channel: Is either private or not a channel");
+            return;
+        }
+        const userId = (await this.main.datastore.getTeam(teamId))!.user_id;
+        const {user} = (await client.users.info({ user: userId })) as UsersInfoResponse;
+        await client.chat.postEphemeral({
+            user: creator,
+            text: `Hint: To bridge to Matrix, run the \`/invite @${user!.name}\` command in this channel.`,
+            channel: channelId,
+        });
+
+        // Create the room first.
+        try {
+            const roomId = await this.createRoomForChannel(teamId, creator, channel);
+            await this.main.actionLink({
+                matrix_room_id: roomId,
+                slack_channel_id: channelId,
+                team_id: teamId,
+            });
+        } catch (ex) {
+            log.error("Failed to provision new room dynamically:", ex);
+        }
+    }
+
+    private async syncChannel(teamId: string, channelItem: ConversationsInfo) {
+        if (!this.shouldTeamSync(teamId, "channel", channelItem.id)) {
+            log.info(`Not syncing`);
+            return;
+        }
+        const client = await this.main.clientFactory.getTeamClient(teamId);
+        const existingChannel = this.main.rooms.getBySlackChannelId(channelItem.id);
+        if (existingChannel) {
+            log.debug("Channel already exists in datastore, not bridging");
+            return;
+        }
+        if (!channelItem.is_channel || channelItem.is_private) {
+            log.warn("Not creating channel: Is either private or not a channel");
+            return;
+        }
+        log.info(`Attempting to dynamically bridge ${channelItem.id} ${channelItem.name}`);
+        const userId = (await this.main.datastore.getTeam(teamId))!.user_id;
+        const {user} = (await client.users.info({ user: userId })) as UsersInfoResponse;
+        try {
+            await client.chat.postEphemeral({
```

Comment:
You shouldn't repeat this, you should have a method to handle it, because then you can have it also do different things.